## Подготовка

### Импортируем необходимые библиотеки и компоненты

In [7]:
import pathlib as pl
import warnings
import re
import random
import requests
import zipfile
import tabulate

from corus import load_ne5, load_lenta
from datasets import Dataset, ClassLabel, Sequence, concatenate_datasets
from natasha import Segmenter, NewsEmbedding, NewsNERTagger, Doc

import evaluate
import torch
import numpy as np
import pandas as pd

from transformers import (
    AutoTokenizer,
    AutoModelForTokenClassification,
    TrainingArguments,
    Trainer,
    DataCollatorForLanguageModeling,
    AutoModelForMaskedLM
)

device = "cuda" if torch.cuda.is_available() else "cpu"

print(device)
warnings.filterwarnings("ignore")

cuda


In [8]:
def convert_markup_to_ner(item):
    text = item.text
    tokens = []
    offsets = []
    for match in re.finditer(r"\S+", text):
        tokens.append(match.group())
        offsets.append((match.start(), match.end()))

    labels = ["O"] * len(tokens)

    for span in item.spans:
        token_indices = [
            i for i, (tstart, tend) in enumerate(offsets) if not (tend <= span.start or tstart >= span.stop)
        ]
        if token_indices:
            labels[token_indices[0]] = "B-" + span.type
            for idx in token_indices[1:]:
                labels[idx] = "I-" + span.type

    return {"id": item.id, "tokens": tokens, "ner_tags": labels}

In [9]:
def convert_labels(example):
    example["ner_tags"] = [label_to_id[label] for label in example["ner_tags"]]
    return example

In [10]:
def compute_metrics(p):
    predictions, labels = p
    predictions = np.argmax(predictions, axis=2)
    true_predictions = [
        [unique_labels[p] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]
    true_labels = [
        [unique_labels[l] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]
    results = metric.compute(predictions=true_predictions, references=true_labels)
    return {
        "precision": results["overall_precision"],
        "recall": results["overall_recall"],
        "f1": results["overall_f1"],
        "accuracy": results["overall_accuracy"],
    }

In [11]:
def join_tokens(example):
    example["text"] = " ".join(example["tokens"])
    return example

In [12]:
def tokenize_and_align_labels(examples):
    tokenized_inputs = tokenizer(
        examples["tokens"],
        truncation=True,
        is_split_into_words=True,
        max_length=128,
        padding="max_length"
    )
    all_labels = []
    for i, labels in enumerate(examples["ner_tags"]):
        word_ids = tokenized_inputs.word_ids(batch_index=i)
        previous_word_idx = None
        label_ids = []
        for word_idx in word_ids:
            if word_idx is None:
                label_ids.append(-100)
            elif word_idx != previous_word_idx:
                label_ids.append(labels[word_idx])
            else:
                label_ids.append(-100)
            previous_word_idx = word_idx
        all_labels.append(label_ids)
    tokenized_inputs["labels"] = all_labels
    return tokenized_inputs

In [13]:
def tokenize_for_mlm(example):
    return tokenizer(example["text"], truncation=True, max_length=512, padding="max_length")

In [14]:
def create_markup_natasha(example):
    text = example["text"]
    doc = Doc(text)
    doc.segment(segmenter)
    doc.tag_ner(ner_tagger)

    tokens = []
    offsets = []
    for match in re.finditer(r"\S+", text):
        tokens.append(match.group())
        offsets.append((match.start(), match.end()))

    labels = ["O"] * len(tokens)

    for span in doc.spans:
        token_indices = [i for i, (s, e) in enumerate(offsets) if not (e <= span.start or s >= span.stop)]
        if token_indices:
            labels[token_indices[0]] = "B-" + span.type
            for idx in token_indices[1:]:
                labels[idx] = "I-" + span.type

    return {"tokens": tokens, "ner_tags": labels, "text": text}

In [15]:
def convert_tags(example):
    example["ner_tags"] = [label_map[tag] for tag in example["ner_tags"]]
    return example

In [16]:
def compute_metrics_aug(p):
    predictions, labels = p
    predictions = np.argmax(predictions, axis=2)

    true_predictions = []
    true_labels = []

    for pred, lab in zip(predictions, labels):
        pred_labels = []
        lab_labels = []
        for p_val, l_val in zip(pred, lab):
            if l_val != -100:
                pred_labels.append(label_list[p_val])
                lab_labels.append(label_list[l_val])
        true_predictions.append(pred_labels)
        true_labels.append(lab_labels)

    results = metric.compute(predictions=true_predictions, references=true_labels)
    return {
        "precision": results["overall_precision"],
        "recall": results["overall_recall"],
        "f1": results["overall_f1"],
        "accuracy": results["overall_accuracy"],
    }

In [17]:
rubert_tiny2_name = "cointegrated/rubert-tiny2"

### Фиксируем seed'ы

In [18]:
SEED = 42
random.seed(SEED)
np.random.seed(SEED)

torch.manual_seed(SEED)
if torch.cuda.is_available():
    torch.cuda.manual_seed(SEED)
    torch.cuda.manual_seed_all(SEED)

torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

## Этап 1 - Загрузка данных

In [19]:
dir_name = "Collection5"
path = f"{dir_name}.zip"

In [20]:
zip_path = pl.Path.cwd()/path

In [21]:
url = "http://www.labinform.ru/pub/named_entities/collection5.zip"

if not pl.Path.exists(zip_path):
    print(f"Файл {path} не найден. Начинаю загрузку...")

    response = requests.get(url, stream=True)
    if response.status_code == 200:
        with open(path, "wb") as file:
            for chunk in response.iter_content(chunk_size=8192):
                file.write(chunk)
        print(f"Файл {path} успешно загружен.")
    else:
        print(f"Ошибка при загрузке файла. Код статуса: {response.status_code}")
else:
    print(f"Файл {path} уже существует. Загрузка не требуется.")

Файл Collection5.zip не найден. Начинаю загрузку...
Файл Collection5.zip успешно загружен.


In [22]:
extract_dir = pl.Path.cwd() # текущая директория
if pl.Path.exists(zip_path):
    print(f"Распаковка ZIP-архива {path}...")

    with zipfile.ZipFile(path, 'r') as zip_ref:
        zip_ref.extractall(extract_dir)

    if not pl.Path.exists(extract_dir):
        print(f"Создаём директорию {extract_dir}...")
        pl.Path.mkdir(extract_dir, exist_ok=True)

    print(f"Архив {path} успешно распакован в директорию {extract_dir}.")
    print(f"Список распакованных файлов: {[p.name for p in pl.Path(pl.Path.cwd()/dir_name).iterdir()]}")
else:
    print(f"Файл {path} не существует.")

Распаковка ZIP-архива Collection5.zip...
Архив Collection5.zip успешно распакован в директорию /content.
Список распакованных файлов: ['053.ann', '510.txt', '1184.ann', 'ryadovoy chelah.ann', 'last_21.txt', '357.ann', 'last_57.ann', '1029.ann', '052.txt', '234.ann', '623.txt', '2031.txt', '09_01_13e.txt', '248.ann', '444.txt', '017.ann', '1115.ann', '596.txt', '090.txt', '160.txt', '587.ann', '1163.txt', '1008.txt', '2010.txt', '04_12_12d.ann', '131.txt', '170.ann', '333.txt', '474.ann', '067.txt', '576.ann', '15_01_13f.ann', '618.ann', '450.ann', 'last_52.txt', '2005.ann', '1168.ann', '134.ann', '086.txt', '303.ann', '570.txt', '585.txt', '388.txt', '464.txt', '528.txt', '300.ann', '630.ann', '349.ann', '572.ann', '128.ann', '1199.txt', '016.ann', '125.ann', '538.ann', '2049.ann', '1102.ann', '615.txt', 'si_tzjanpin.ann', 'last_29.txt', '564.ann', '373.ann', '501.ann', '1147.txt', '289.txt', '1136.txt', '270.txt', '493.ann', '21_11_12i.ann', '536.txt', '1162.txt', '22_11_12c.ann', '28

In [23]:
dataset = load_ne5("Collection5")

In [24]:
next(dataset)

Ne5Markup(
    id='510',
    text='Д.Медведев снял с должности замсекретаря Совбеза РФ Ю.Балуевского\r\n\r\nПрезидент России Дмитрий Медведев освободил Юрия Балуевского от должности замсекретаря Совета безопасности России. Соответствующий указ опубликован на сайте государственной системы правовой информации.\r\n\r\nУказ глава государства подписал 9 января. Именно в этот день Ю.Балуевскому исполнилось 65 лет.\r\n\r\nСогласно российскому законодательству, предельный возраст для нахождения на государственной службе составляет 60 лет, однако он может быть увеличен до 65 лет. Дальнейшее продление нахождения на госслужбе не допускается.\r\n\r\nЮ.Балуевский родился 9 января 1947г. в городе Трускавец Львовской области Украины. В 1970г. окончил Ленинградское общевойсковое училище, в 1980г. - Военную академию имени Фрунзе, в 1990г. - Военную академию генштаба Вооруженных сил СССР.\r\n\r\n19 июля 2004г. указом президента Ю.Балуевский был назначен начальником Генерального штаба Вооруженных сил РФ 

### Минимальная предобработка данных

* Будем решать задачу как задачу классификации - набор токенов к метке.
* Для этого предобработаем текст до вида list токенов, list меток.
* Метки представим по схеме BIO.

In [25]:
data_list = [convert_markup_to_ner(item) for item in dataset]

In [26]:
ner_dataset = Dataset.from_list(data_list)

In [27]:
unique_labels = set()
for example in ner_dataset:
    unique_labels.update(example["ner_tags"])

unique_labels = sorted(list(unique_labels))
label_to_id = {label: i for i, label in enumerate(unique_labels)}

In [28]:
label_to_id

{'B-GEOPOLIT': 0,
 'B-LOC': 1,
 'B-MEDIA': 2,
 'B-ORG': 3,
 'B-PER': 4,
 'I-GEOPOLIT': 5,
 'I-LOC': 6,
 'I-MEDIA': 7,
 'I-ORG': 8,
 'I-PER': 9,
 'O': 10}

In [29]:
ner_dataset = ner_dataset.map(convert_labels)

Map:   0%|          | 0/999 [00:00<?, ? examples/s]

In [30]:
features = ner_dataset.features.copy()
features["ner_tags"] = Sequence(ClassLabel(names=unique_labels))
ner_dataset = ner_dataset.cast(features)

Casting the dataset:   0%|          | 0/999 [00:00<?, ? examples/s]

In [31]:
print(ner_dataset[0])

{'id': 'last_21', 'tokens': ['ФОМ:', 'более', '40%', 'жителей', 'Подмосковья', 'готовы', 'голосовать', 'за', 'Воробьева', 'Фонд', '"Общественное', 'мнение"', 'выяснил,', 'за', 'кого', 'собираются', 'голосовать', 'жители', 'Московской', 'области', 'на', 'выборах', 'губернатора', '8', 'сентября.', 'Более', '40%', 'опрошенных', 'Фондом', '"Общественное', 'мнение"', '(ФОМ)', 'респондентов', 'сообщили,', 'что', 'готовы', 'голосовать', 'за', 'Андрея', 'Воробьева', 'на', 'выборах', 'губернатора', 'Московской', 'области,', 'еще', '40%', 'пока', 'не', 'решили,', 'за', 'кого', 'собираются', 'отдать', 'свой', 'голос,', 'сообщил', 'в', 'четверг', 'президент', 'фонда', 'Александр', 'Ослон', 'на', 'пресс-конференции', 'в', 'РИА', 'Новости.', 'Ослон', 'представил', '1', 'августа', 'в', 'РИА', 'Новости', 'интернет-проект', '"Как', 'сделать', 'лучше', 'наше', 'Подмосковье",', 'который', 'позиционируется', 'как', 'новый', 'способ', 'управления', 'регионом', 'вместе', 'с', 'населением.', 'На', 'портале',

In [32]:
tokenizer = AutoTokenizer.from_pretrained(rubert_tiny2_name)

tokenizer_config.json:   0%|          | 0.00/401 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/1.08M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.74M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

### Токенизируем текст

Оставим оригинальные метки для первого токена в последовательности, остальные отметим -100.

In [33]:
tokenized_dataset = ner_dataset.map(tokenize_and_align_labels, batched=True)

Map:   0%|          | 0/999 [00:00<?, ? examples/s]

In [34]:
tokenized_dataset = tokenized_dataset.train_test_split(test_size=0.2)

### Разделим данные на train-test

In [35]:
train_dataset = tokenized_dataset["train"]
test_dataset = tokenized_dataset["test"]

## Этап 2 - Обучение модели на задачу NER

In [36]:
num_labels = len(unique_labels)
model = AutoModelForTokenClassification.from_pretrained(rubert_tiny2_name, num_labels=num_labels)

config.json:   0%|          | 0.00/693 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/118M [00:00<?, ?B/s]

Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


### Используем стандартные гиперпараметры для подобных моделей

In [37]:
training_args = TrainingArguments(
    output_dir="models/dbert",
    eval_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=20,
    weight_decay=0.01,
)

### Используем метрики классификации для sequence labeling

In [38]:
metric = evaluate.load("seqeval")

### Обучим модель

In [39]:
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics,
)

In [40]:
print("Метрики ДО дообучения:")
pre_training_results = trainer.evaluate()
print(tabulate.tabulate(
    pre_training_results.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))

Метрики ДО дообучения:


wandb: WARNING The `run_name` is currently set to the same value as `TrainingArguments.output_dir`. If this was not intended, please specify a different run name by setting the `TrainingArguments.run_name` parameter.
wandb: Using wandb-core as the SDK backend.  Please refer to https://wandb.me/wandb-core for more information.


<IPython.core.display.Javascript object>

wandb: WARNING If you're specifying your api key in code, ensure this code is not shared publicly.
wandb: WARNING Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.
wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: nikolay-tep (nikolay-tep-itmo-university) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     2.6935 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0014 |
+-----------------------------+------------+
| eval_precision              |     0.0170 |
+-----------------------------+------------+
| eval_recall                 |     0.0584 |
+-----------------------------+------------+
| eval_f1                     |     0.0263 |
+-----------------------------+------------+
| eval_accuracy               |     0.0161 |
+-----------------------------+------------+
| eval_runtime                |     1.1678 |
+-----------------------------+------------+
| eval_samples_per_second     |   171.2560 |
+-----------------------------+------------+
| eval_steps_per_second       |    11.1320 |
+-----------------------------+------------+


In [41]:
trainer.train()

Epoch,Training Loss,Validation Loss,Model Preparation Time,Precision,Recall,F1,Accuracy
1,No log,0.928988,0.001400,0.000000,0.000000,0.000000,0.767672
2,No log,0.555164,0.001400,0.263158,0.247678,0.255183,0.834142
3,No log,0.411422,0.001400,0.438282,0.501548,0.467786,0.881728
4,No log,0.337083,0.001400,0.574329,0.645898,0.608015,0.909332
5,No log,0.286173,0.001400,0.634163,0.699690,0.665317,0.923423
6,No log,0.250423,0.001400,0.677925,0.753483,0.713710,0.933356
7,No log,0.219523,0.001400,0.694737,0.766254,0.728745,0.941499
8,No log,0.197384,0.001400,0.707940,0.783282,0.743708,0.948429
9,No log,0.179710,0.001400,0.732086,0.802632,0.765737,0.953222
10,0.467100,0.167947,0.001400,0.755160,0.821207,0.786800,0.957554


TrainOutput(global_step=1000, training_loss=0.30087158966064453, metrics={'train_runtime': 49.5062, 'train_samples_per_second': 322.788, 'train_steps_per_second': 20.199, 'total_flos': 28296031964160.0, 'train_loss': 0.30087158966064453, 'epoch': 20.0})

In [42]:
print("Метрики ПОСЛЕ дообучения:")
post_training_results = trainer.evaluate()
print(tabulate.tabulate(
    post_training_results.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))

Метрики ПОСЛЕ дообучения:


+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     0.1296 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0014 |
+-----------------------------+------------+
| eval_precision              |     0.8130 |
+-----------------------------+------------+
| eval_recall                 |     0.8649 |
+-----------------------------+------------+
| eval_f1                     |     0.8382 |
+-----------------------------+------------+
| eval_accuracy               |     0.9678 |
+-----------------------------+------------+
| eval_runtime                |     0.6232 |
+-----------------------------+------------+
| eval_samples_per_second     |   320.8990 |
+-----------------------------+------------+
| eval_steps_per_second       |    20.8580 |
+-----------------------------+------------+
| epoch                       |    20.0000 |
+---------

### Выводы по эксперименту

* Метрики после обучения значительно лучше, чем до, что очевидно.
* Получили очень высокий `accuracy` ≈ 0.96, что понятно по метке 'O' - пустой, их большинство.
* Получили вполне неплохой `f1` ≈ 0.8382 на тесте, будем в дальнейшем сравнивать этот показатель.

## Этап 3 - Предобучим на MLM

### Подготовим данные

Склеим тексты, токенизируем их для нашей модели.

In [43]:
mlm_train_dataset = train_dataset.map(join_tokens)
mlm_test_dataset = test_dataset.map(join_tokens)

Map:   0%|          | 0/799 [00:00<?, ? examples/s]

Map:   0%|          | 0/200 [00:00<?, ? examples/s]

In [44]:
mlm_train_dataset = mlm_train_dataset.map(tokenize_for_mlm, batched=True)
mlm_train_dataset.set_format(type="torch", columns=["input_ids", "attention_mask"])

mlm_test_dataset = mlm_test_dataset.map(tokenize_for_mlm, batched=True)
mlm_test_dataset.set_format(type="torch", columns=["input_ids", "attention_mask"])

Map:   0%|          | 0/799 [00:00<?, ? examples/s]

Map:   0%|          | 0/200 [00:00<?, ? examples/s]

### Создадим модель и обучим её

Возьмём меньшее количество эпох, но чуть повысим LR.

In [45]:
mlm_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=True, mlm_probability=0.15)
mlm_model = AutoModelForMaskedLM.from_pretrained(rubert_tiny2_name)

In [46]:
mlm_training_args = TrainingArguments(
    output_dir="models/mlm/",
    num_train_epochs=5,
    per_device_train_batch_size=16,
    eval_strategy="epoch",
    learning_rate=5e-5,
    weight_decay=0.01,
)

mlm_trainer = Trainer(
    model=mlm_model,
    args=mlm_training_args,
    train_dataset=mlm_train_dataset,
    eval_dataset=mlm_test_dataset,
    data_collator=mlm_collator,
)

In [47]:
print("Метрики ДО дообучения MLM:")
pre_training_mlm_results = mlm_trainer.evaluate()
print(tabulate.tabulate(
    pre_training_mlm_results.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))

Метрики ДО дообучения MLM:


+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     2.9991 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0009 |
+-----------------------------+------------+
| eval_runtime                |     2.4713 |
+-----------------------------+------------+
| eval_samples_per_second     |    80.9280 |
+-----------------------------+------------+
| eval_steps_per_second       |    10.1160 |
+-----------------------------+------------+


In [48]:
mlm_trainer.train()

Epoch,Training Loss,Validation Loss,Model Preparation Time
1,No log,2.899897,0.000900
2,No log,2.818942,0.000900
3,No log,2.884795,0.000900
4,No log,2.858882,0.000900
5,No log,2.873257,0.000900


TrainOutput(global_step=250, training_loss=3.08001806640625, metrics={'train_runtime': 177.5078, 'train_samples_per_second': 22.506, 'train_steps_per_second': 1.408, 'total_flos': 30488723189760.0, 'train_loss': 3.08001806640625, 'epoch': 5.0})

In [49]:
print("Метрики ПОСЛЕ дообучения MLM:")
post_training_mlm_results = mlm_trainer.evaluate()
print(tabulate.tabulate(
    post_training_mlm_results.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))

Метрики ПОСЛЕ дообучения MLM:


+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     2.8244 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0009 |
+-----------------------------+------------+
| eval_runtime                |     2.4215 |
+-----------------------------+------------+
| eval_samples_per_second     |    82.5920 |
+-----------------------------+------------+
| eval_steps_per_second       |    10.3240 |
+-----------------------------+------------+
| epoch                       |     5.0000 |
+-----------------------------+------------+


### Промежуточный вывод

Успешно обучили модель MLM, проверим на NER.

### Обучаем NER модель с момента чекпоинта MLM

Оставим все оригинальные параметры. Возможно, это не оптимальный подход, но можно будет объективно сравнить.

In [50]:
model = AutoModelForTokenClassification.from_pretrained("models/mlm/checkpoint-250/", num_labels=num_labels)

Some weights of BertForTokenClassification were not initialized from the model checkpoint at models/mlm/checkpoint-250/ and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [51]:
ner_training_args = TrainingArguments(
    output_dir="models/mlm/",
    eval_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=20,
    weight_decay=0.01,
)

ner_trainer = Trainer(
    model=model,
    args=ner_training_args,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics,
)

In [52]:
print("Метрики ДО дообучения NER:")
pre_training_ner_results = ner_trainer.evaluate()
print(tabulate.tabulate(
    pre_training_ner_results.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))

Метрики ДО дообучения NER:


+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     2.4171 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0010 |
+-----------------------------+------------+
| eval_precision              |     0.0158 |
+-----------------------------+------------+
| eval_recall                 |     0.0820 |
+-----------------------------+------------+
| eval_f1                     |     0.0265 |
+-----------------------------+------------+
| eval_accuracy               |     0.0900 |
+-----------------------------+------------+
| eval_runtime                |     0.4384 |
+-----------------------------+------------+
| eval_samples_per_second     |   456.2260 |
+-----------------------------+------------+
| eval_steps_per_second       |    29.6550 |
+-----------------------------+------------+


In [53]:
ner_trainer.train()

Epoch,Training Loss,Validation Loss,Model Preparation Time,Precision,Recall,F1,Accuracy
1,No log,0.905190,0.001000,0.000000,0.000000,0.000000,0.767672
2,No log,0.537608,0.001000,0.232034,0.234907,0.233462,0.834777
3,No log,0.391986,0.001000,0.413850,0.483359,0.445912,0.882363
4,No log,0.311843,0.001000,0.569703,0.638932,0.602335,0.916436
5,No log,0.254269,0.001000,0.682848,0.734907,0.707922,0.937688
6,No log,0.212819,0.001000,0.723604,0.777090,0.749394,0.947967
7,No log,0.184730,0.001000,0.749456,0.799923,0.773867,0.954262
8,No log,0.167389,0.001000,0.767735,0.825077,0.795374,0.958709
9,No log,0.153570,0.001000,0.802776,0.850619,0.826005,0.963964
10,0.432100,0.143292,0.001000,0.815046,0.859520,0.836692,0.966101


TrainOutput(global_step=1000, training_loss=0.2750410270690918, metrics={'train_runtime': 42.6988, 'train_samples_per_second': 374.249, 'train_steps_per_second': 23.42, 'total_flos': 28296031964160.0, 'train_loss': 0.2750410270690918, 'epoch': 20.0})

In [54]:
print("Метрики ПОСЛЕ дообучения NER:")
post_training_ner_results = ner_trainer.evaluate()
print(tabulate.tabulate(
    post_training_ner_results.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))

Метрики ПОСЛЕ дообучения NER:


+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     0.1141 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0010 |
+-----------------------------+------------+
| eval_precision              |     0.8392 |
+-----------------------------+------------+
| eval_recall                 |     0.8827 |
+-----------------------------+------------+
| eval_f1                     |     0.8604 |
+-----------------------------+------------+
| eval_accuracy               |     0.9716 |
+-----------------------------+------------+
| eval_runtime                |     0.3565 |
+-----------------------------+------------+
| eval_samples_per_second     |   560.9640 |
+-----------------------------+------------+
| eval_steps_per_second       |    36.4630 |
+-----------------------------+------------+
| epoch                       |    20.0000 |
+---------

### Выводы по эксперименту
* Возможно, слегка недообучили модель (видно по валидационной выборке)
* Получили качество чуть лучше по `f1` ≈ 0.8604 и `accuracy` ≈ 0.9716, но, вполне вероятно, что при полном обучении обеих моделей, результаты были бы +- схожие.

## Этап 4 - Аугментация с помощью большой модели NER

* Используем 10_000 сэмплов lenta
* В качестве модели возьмём NewsNERTagger из natasha, он имеет хорошее качество по бенчмаркам при невероятной по сравнению с конкурентами производительности.

### Аналогично предобработаем данные

In [55]:
!wget https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz

--2025-04-15 09:20:58--  https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250415%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250415T092058Z&X-Amz-Expires=300&X-Amz-Signature=bbe1fa920e30a0fa365fa86d2e44e5df18b52bc0cae217e58867542b2b6a2965&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dlenta-ru-news.csv.gz&response-content-type=application%2Foctet-stream [following]
--2025-04-15 09:20:58--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-H

In [57]:
data_lenta = load_lenta("lenta-ru-news.csv.gz")

target_columns = ["title", "topic", "text"]
data_dict = {c: [] for c in target_columns}

for item in data_lenta:
    for column in target_columns:
        data_dict[column].append(eval(f"item.{column}"))

df = pd.DataFrame(data_dict)
df = df.sample(10_000, random_state=SEED)

dataset = Dataset.from_pandas(df)

### Используем намеченный natasha-сетап

In [60]:
segmenter = Segmenter()
embedding = NewsEmbedding()
ner_tagger = NewsNERTagger(embedding)

In [61]:
markup_dataset = dataset.map(create_markup_natasha)

Map:   0%|          | 0/10000 [00:00<?, ? examples/s]

### Приведём лейблы к виду нашего train датасета для объединения

In [62]:
label_list = [
    "B-GEOPOLIT",
    "B-LOC",
    "B-MEDIA",
    "B-ORG",
    "B-PER",
    "I-GEOPOLIT",
    "I-LOC",
    "I-MEDIA",
    "I-ORG",
    "I-PER",
    "O",
]
label_map = {label: i for i, label in enumerate(label_list)}

markup_dataset = markup_dataset.map(convert_tags)

Map:   0%|          | 0/10000 [00:00<?, ? examples/s]

In [63]:
features = markup_dataset.features.copy()
features["ner_tags"] = Sequence(feature=ClassLabel(names=label_list))
markup_dataset = markup_dataset.cast(features)

Casting the dataset:   0%|          | 0/10000 [00:00<?, ? examples/s]

In [64]:
markup_dataset = markup_dataset.map(tokenize_and_align_labels, batched=True)
markup_dataset = markup_dataset.remove_columns(["title", "topic", "text"])

Map:   0%|          | 0/10000 [00:00<?, ? examples/s]

In [65]:
combined_train_dataset = concatenate_datasets([train_dataset, markup_dataset])

### Обучим модель

Для чистоты проведения эксперимента, оставим параметры и метрики аналогичными предыдущим.

In [66]:
unique_labels = set()
for example in combined_train_dataset:
    unique_labels.update(example["ner_tags"])
unique_labels = sorted(list(unique_labels))
unique_labels

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [67]:
model = AutoModelForTokenClassification.from_pretrained(rubert_tiny2_name, num_labels=len(unique_labels))

Some weights of BertForTokenClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [68]:
aug_training_args = TrainingArguments(
    output_dir="models/augmented/",
    eval_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=20,
    weight_decay=0.01,
)

aug_trainer = Trainer(
    model=model,
    args=aug_training_args,
    train_dataset=combined_train_dataset,
    eval_dataset=test_dataset,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics_aug,
)

In [69]:
print("Метрики ДО дообучения NER:")
pre_training_aug_results = aug_trainer.evaluate()
print(tabulate.tabulate(
    pre_training_aug_results.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))

Метрики ДО дообучения NER:


+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     2.3433 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0011 |
+-----------------------------+------------+
| eval_precision              |     0.0244 |
+-----------------------------+------------+
| eval_recall                 |     0.1215 |
+-----------------------------+------------+
| eval_f1                     |     0.0406 |
+-----------------------------+------------+
| eval_accuracy               |     0.1441 |
+-----------------------------+------------+
| eval_runtime                |     0.4358 |
+-----------------------------+------------+
| eval_samples_per_second     |   458.9020 |
+-----------------------------+------------+
| eval_steps_per_second       |    29.8290 |
+-----------------------------+------------+


In [70]:
aug_trainer.train()

Epoch,Training Loss,Validation Loss,Model Preparation Time,Precision,Recall,F1,Accuracy
1,0.340400,0.267280,0.001100,0.655237,0.672988,0.663994,0.933356
2,0.098900,0.215294,0.001100,0.677908,0.692337,0.685047,0.939074
3,0.057000,0.171647,0.001100,0.695833,0.710913,0.703292,0.944560
4,0.046900,0.149638,0.001100,0.738473,0.756192,0.747228,0.951432
5,0.043500,0.149034,0.001100,0.746662,0.757353,0.751969,0.952703
6,0.033600,0.139509,0.001100,0.769669,0.783669,0.776606,0.956803
7,0.030500,0.129741,0.001100,0.781025,0.796440,0.788657,0.959402
8,0.027100,0.115574,0.001100,0.793807,0.813467,0.803517,0.963155
9,0.024100,0.114938,0.001100,0.799773,0.819272,0.809405,0.964137
10,0.021500,0.118214,0.001100,0.799168,0.817724,0.808340,0.963848


TrainOutput(global_step=13500, training_loss=0.03933262069137008, metrics={'train_runtime': 618.9621, 'train_samples_per_second': 348.939, 'train_steps_per_second': 21.811, 'total_flos': 382439110364160.0, 'train_loss': 0.03933262069137008, 'epoch': 20.0})

In [71]:
print("Метрики ПОСЛЕ дообучения NER:")
pre_training_aug_results = aug_trainer.evaluate()
print(tabulate.tabulate(
    pre_training_aug_results.items(),
    headers=["Метрика", "Значение"],
    tablefmt="grid",
    floatfmt=".4f"
))

Метрики ПОСЛЕ дообучения NER:


+-----------------------------+------------+
| Метрика                     |   Значение |
+=============================+============+
| eval_loss                   |     0.1093 |
+-----------------------------+------------+
| eval_model_preparation_time |     0.0011 |
+-----------------------------+------------+
| eval_precision              |     0.8284 |
+-----------------------------+------------+
| eval_recall                 |     0.8483 |
+-----------------------------+------------+
| eval_f1                     |     0.8382 |
+-----------------------------+------------+
| eval_accuracy               |     0.9695 |
+-----------------------------+------------+
| eval_runtime                |     0.3426 |
+-----------------------------+------------+
| eval_samples_per_second     |   583.7390 |
+-----------------------------+------------+
| eval_steps_per_second       |    37.9430 |
+-----------------------------+------------+
| epoch                       |    20.0000 |
+---------

### Выводы по эксперименту

* Возможно слегка недообучили модель (видно по валидационной выборке)
* Для текущих данных метрики качества получились выше, чем у базовой модели, но ниже чем с `MLM`
* `f1` = 0.8382

## Финальные выводы

* Смог обучить весьма неплохую модель `f1` ≈ 0.8604.
* MLM-дообучение + NER и Синтетическая разметка + NER показали улучшение результатов.
* Предобучение MLM в данной задаче оказалось полезным.
* При других данных или специфике текста итоговые результаты могли бы отличаться.